File: packages/pty-testing/src/keys.ts
/**
 * Key names to the bytes a terminal sends.
 *
 * `ctrl+u`, `ctrl-u`, `ctrl_u` and `C-u` all mean the same key. The
 * modifiers are ctrl, alt and shift; the keys are a to z and the names
 * below.
 */

const KEY_MAP: Record<string, string> = {
  return: "\r",
  enter: "\r",
  tab: "\t",
  escape: "\x1b",
  esc: "\x1b",
  space: " ",
  backspace: "\x7f",
  delete: "\x1b[3~",
  up: "\x1b[A",
  down: "\x1b[B",
  right: "\x1b[C",
  left: "\x1b[D",
  home: "\x1b[H",
  end: "\x1b[F",
  pageup: "\x1b[5~",
  pagedown: "\x1b[6~",
};

const MODIFIERS = new Set(["ctrl", "alt", "shift"]);
const SEPARATORS = /[+_-]/;
const NAMED_KEYS = Object.keys(KEY_MAP).sort().join(", ");

const HELP =
  `Use ctrl+u, ctrl-u, ctrl_u, or C-u; supported modifiers are ctrl, alt, and shift; ` +
  `supported keys are a-z, ${NAMED_KEYS}.`;

/** Keycodes for the control keys under a modifier (CSI u). */
const CSI_U: Record<string, number> = {
  return: 13,
  enter: 13,
  tab: 9,
  escape: 27,
  esc: 27,
  space: 32,
  backspace: 127,
};

/** 1 + shift(1) + alt(2) + ctrl(4), the xterm modifier parameter. */
function modifierParam(mods: Set<string>): number {
  return (
    1 + (mods.has("shift") ? 1 : 0) + (mods.has("alt") ? 2 : 0) + (mods.has("ctrl") ? 4 : 0)
  );
}

/**
 * `C-u` is the readline and tmux spelling of ctrl+u. The one-letter alias is
 * scoped to a leading `C-`, so `C+u` keeps no surprise meaning.
 */
function normalizeModifier(mod: string, index: number, spec: string): string {
  if (mod === "c" && index === 0 && /^c-/i.test(spec)) return "ctrl";
  return mod;
}

function isSupportedBase(base: string): boolean {
  return KEY_MAP[base] !== undefined || (base.length === 1 && base >= "a" && base <= "z");
}

/** The bytes for a key spec. Throws with the help text on a bad one. */
export function resolveKey(spec: string): string {
  const normalized = spec.toLowerCase();
  const hasSeparator = SEPARATORS.test(normalized);
  const rawParts = hasSeparator ? normalized.split(SEPARATORS) : [normalized];
  const rawBase = rawParts.at(-1) ?? "";
  const rawMods = rawParts.slice(0, -1).map((m, i) => normalizeModifier(m, i, spec));

  // A name with a separator in it could be read two ways. Refuse rather than
  // silently pick one.
  const isValidChord =
    rawBase !== "" &&
    rawMods.length > 0 &&
    rawMods.every((m) => m !== "" && MODIFIERS.has(m)) &&
    isSupportedBase(rawBase);
  if (hasSeparator && KEY_MAP[normalized] !== undefined && isValidChord) {
    throw new Error(
      `Ambiguous key spec "${spec}": it is both a named key and a modifier chord. ${HELP}`,
    );
  }
  if (KEY_MAP[normalized] !== undefined && !isValidChord) return KEY_MAP[normalized];

  const parts = [...rawParts];
  const base = parts.pop() as string;
  if (base === "" || parts.some((p) => p === "")) {
    throw new Error(`Incomplete key spec "${spec}". ${HELP}`);
  }
  const mods = new Set(parts.map((m, i) => normalizeModifier(m, i, spec)));
  for (const mod of mods) {
    if (!MODIFIERS.has(mod)) {
      throw new Error(`Unknown modifier: "${mod}" in key spec "${spec}". ${HELP}`);
    }
  }

  const isLetter = base.length === 1 && base >= "a" && base <= "z";
  const mapped = KEY_MAP[base];
  if (mapped === undefined && !isLetter) {
    throw new Error(`Unknown key: "${base}" in key spec "${spec}". ${HELP}`);
  }

  if (isLetter) {
    let result = base;
    if (mods.has("shift")) result = result.toUpperCase();
    if (mods.has("ctrl")) result = String.fromCharCode(result.toLowerCase().charCodeAt(0) - 96);
    if (mods.has("alt")) result = "\x1b" + result;
    return result;
  }

  if (mods.size === 0) return mapped;

  const param = modifierParam(mods);
  // shift+tab has its own old sequence.
  if (base === "tab" && param === 2) return "\x1b[Z";

  const tilde = /^\x1b\[(\d+)~$/.exec(mapped);
  if (tilde) return `\x1b[${tilde[1]};${param}~`;
  const letter = /^\x1b\[([A-Z])$/.exec(mapped);
  if (letter) return `\x1b[1;${param}${letter[1]}`;
  const code = CSI_U[base];
  if (code !== undefined) return `\x1b[${code};${param}u`;
  return mapped;
}

File: crates/pty-core/src/keys.rs
//! Named-key resolution, ported from the pty project's `src/keys.ts`.
//!
//! Turns key specs like `ctrl+c`, `return`, `alt+x`, `shift+tab` into the byte
//! sequence a real terminal would send to the PTY. Modifier combinations use
//! the xterm modifier-parameter encoding; control chars use CSI-u.

use std::collections::HashSet;

/// Error returned when a key spec references an unknown key or modifier.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct KeyError(pub String);

impl std::fmt::Display for KeyError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}", self.0)
    }
}

impl std::error::Error for KeyError {}

/// Base named keys → their unmodified byte sequence.
fn key_map(name: &str) -> Option<&'static str> {
    Some(match name {
        "return" | "enter" => "\r",
        "tab" => "\t",
        "escape" | "esc" => "\x1b",
        "space" => " ",
        "backspace" => "\x7f",
        "delete" => "\x1b[3~",
        "up" => "\x1b[A",
        "down" => "\x1b[B",
        "right" => "\x1b[C",
        "left" => "\x1b[D",
        "home" => "\x1b[H",
        "end" => "\x1b[F",
        "pageup" => "\x1b[5~",
        "pagedown" => "\x1b[6~",
        _ => return None,
    })
}

/// CSI-u keycodes for control-char keys under modifiers.
fn csi_u_keycode(name: &str) -> Option<u32> {
    Some(match name {
        "return" | "enter" => 13,
        "tab" => 9,
        "escape" | "esc" => 27,
        "space" => 32,
        "backspace" => 127,
        _ => return None,
    })
}

fn is_modifier(m: &str) -> bool {
    matches!(m, "ctrl" | "alt" | "shift")
}

/// The named keys, sorted, as the help text lists them.
fn named_keys() -> String {
    let mut names = [
        "return", "enter", "tab", "escape", "esc", "space", "backspace", "delete", "up", "down",
        "right", "left", "home", "end", "pageup", "pagedown",
    ];
    names.sort_unstable();
    names.join(", ")
}

/// The sentence every key-spec error ends with.
///
/// node: src/keys.ts:22-25 (`KEY_SPEC_HELP`)
fn key_spec_help() -> String {
    format!(
        "Use ctrl+u, ctrl-u, ctrl_u, or C-u; supported modifiers are ctrl, alt, and shift; \
         supported keys are a-z, {}.",
        named_keys()
    )
}

/// `+`, `_` and `-` all separate a modifier from what follows it.
///
/// node: src/keys.ts:21 (`MODIFIER_SEPARATORS`)
fn is_separator(c: char) -> bool {
    c == '+' || c == '_' || c == '-'
}

/// `C-u` is the readline and tmux spelling of `ctrl+u`. The one-letter alias
/// is scoped to a leading `C-`, so `C+u` keeps no surprise meaning.
///
/// node: src/keys.ts:47-53 (`normalizeModifier`)
fn normalize_modifier(m: &str, index: usize, spec: &str) -> String {
    let leading_c_dash = {
        let mut chars = spec.chars();
        matches!(chars.next(), Some(c) if c.eq_ignore_ascii_case(&'c'))
            && matches!(chars.next(), Some('-'))
    };
    if m == "c" && index == 0 && leading_c_dash {
        return "ctrl".to_string();
    }
    m.to_string()
}

fn is_supported_base(base: &str) -> bool {
    key_map(base).is_some() || (base.len() == 1 && base.as_bytes()[0].is_ascii_lowercase())
}

/// xterm modifier parameter: `1 + bitmask(shift=1, alt=2, ctrl=4)`.
fn modifier_param(mods: &HashSet<String>) -> u32 {
    1 + if mods.contains("shift") { 1 } else { 0 }
        + if mods.contains("alt") { 2 } else { 0 }
        + if mods.contains("ctrl") { 4 } else { 0 }
}

/// Parse a key spec like `ctrl+c`, `return`, `alt+x` into the bytes a terminal
/// would send. Returns `Err` for unknown modifiers or keys.
pub fn resolve_key(spec: &str) -> Result<String, KeyError> {
    let lower = spec.to_lowercase();
    let has_separator = lower.chars().any(is_separator);
    let raw_parts: Vec<&str> = if has_separator {
        lower.split(is_separator).collect()
    } else {
        vec![lower.as_str()]
    };
    let raw_base = *raw_parts.last().unwrap_or(&"");
    let raw_mods: Vec<String> = raw_parts[..raw_parts.len().saturating_sub(1)]
        .iter()
        .enumerate()
        .map(|(i, m)| normalize_modifier(m, i, spec))
        .collect();

    // A separator-bearing name could be both a named key and a modifier
    // chord. Refuse the collision rather than silently pick one.
    //
    // node: src/keys.ts:73-82
    let is_valid_chord = !raw_base.is_empty()
        && !raw_mods.is_empty()
        && raw_mods.iter().all(|m| !m.is_empty() && is_modifier(m))
        && is_supported_base(raw_base);
    if has_separator && key_map(&lower).is_some() && is_valid_chord {
        return Err(KeyError(format!(
            "Ambiguous key spec \"{spec}\": it is both a named key and a modifier chord. {}",
            key_spec_help()
        )));
    }
    if let Some(mapped) = key_map(&lower)
        && !is_valid_chord
    {
        return Ok(mapped.to_string());
    }

    let mut parts = raw_parts;
    let base = parts.pop().unwrap_or("").to_string();
    if base.is_empty() || parts.iter().any(|p| p.is_empty()) {
        return Err(KeyError(format!(
            "Incomplete key spec \"{spec}\". {}",
            key_spec_help()
        )));
    }
    let mods: HashSet<String> = parts
        .iter()
        .enumerate()
        .map(|(i, m)| normalize_modifier(m, i, spec))
        .collect();

    for m in &mods {
        if !is_modifier(m) {
            return Err(KeyError(format!(
                "Unknown modifier: \"{m}\" in key spec \"{spec}\". {}",
                key_spec_help()
            )));
        }
    }

    let is_letter = base.len() == 1 && base.as_bytes()[0].is_ascii_lowercase();
    let has_modifiers = !mods.is_empty();
    let mapped = key_map(&base);

    if mapped.is_none() && !is_letter {
        return Err(KeyError(format!(
            "Unknown key: \"{base}\" in key spec \"{spec}\". {}",
            key_spec_help()
        )));
    }

    // Single-letter keys.
    if is_letter {
        let mut result = base.clone();
        if mods.contains("shift") {
            result = result.to_uppercase();
        }
        if mods.contains("ctrl") {
            let code = result.to_lowercase().as_bytes()[0] as u32;
            result = char::from_u32(code - 96).unwrap().to_string();
        }
        if mods.contains("alt") {
            result = format!("\x1b{result}");
        }
        return Ok(result);
    }

    let mapped = mapped.unwrap();

    // Named keys without modifiers: return the mapped value directly.
    if !has_modifiers {
        return Ok(mapped.to_string());
    }

    let modp = modifier_param(&mods);

    // Special case: shift+tab produces the legacy backtab sequence.
    if base == "tab" && modp == 2 {
        return Ok("\x1b[Z".to_string());
    }

    // CSI sequences of form ESC [ N ~  (e.g. delete, pageup).
    if let Some(digits) = mapped
        .strip_prefix("\x1b[")
        .and_then(|s| s.strip_suffix('~'))
        && !digits.is_empty() && digits.bytes().all(|b| b.is_ascii_digit()) {
            return Ok(format!("\x1b[{digits};{modp}~"));
        }

    // CSI sequences of form ESC [ X  (single uppercase letter: arrows, home, end).
    if let Some(letter) = mapped.strip_prefix("\x1b[")
        && letter.len() == 1 && letter.as_bytes()[0].is_ascii_uppercase() {
            return Ok(format!("\x1b[1;{modp}{letter}"));
        }

    // Control-char keys (return, tab, escape, space, backspace): CSI-u encoding.
    if let Some(keycode) = csi_u_keycode(&base) {
        return Ok(format!("\x1b[{keycode};{modp}u"));
    }

    Ok(mapped.to_string())
}

/// If `value` starts with `key:`, resolve the key name; otherwise return the
/// literal string.
pub fn parse_seq_value(value: &str) -> Result<String, KeyError> {
    if let Some(rest) = value.strip_prefix("key:") {
        resolve_key(rest)
    } else {
        Ok(value.to_string())
    }
}
